Add "emulator" function to systems

matthewbauer opened the pull request:
exeWrapper is based on meson's exe_wrapper idea. Basically, it will give you an emulator that can run whatever executable you want. It is useful for instance in "waf" build systems that expect a "--cross-execute" flag to run some configuration tests. We can use QEMU to emulate Linxux systems. On windows systems we can use WINE.

I am hoping this is useful in more things, for instance running, some tests in the emulator when doing cross compilation.

The QEMU system should be able to run Linux. Running ```pkgs.exeWrapper { config = "aarch64-unknown-linux-gnu"; }```, will give you a bash script to emulate an aarch64 system. I have already rewritten the Nixpkgs cross tests to use this (and can probably add them to release-cross.nix).

GrahamcOfBorg commented:
<!--REQUEST_ID=c003f65a-8fec-404e-9930-2f95f06e73a1-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=6153ffeb-238d-444a-857d-e72e05ac3d95)

Attempted: libxml2, talloc, unscd

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53
shrinking /nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53/bin/nscd
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53/bin
patching script interpreter paths in /nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53
checking for references to /build in /nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53...
/nix/store/h73hfs1kkr9r8zr8rgjzivdc2ff0azdq-libxml2-2.9.8-bin
/nix/store/h50z2qsy9mpl3swqp5pzwp0q6nblc83a-talloc-2.1.14
/nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=c003f65a-8fec-404e-9930-2f95f06e73a1-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=a630c910-a012-40de-90cf-af4aaeaaf69d)

Attempted: libxml2, talloc, unscd

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14/lib/libtalloc.so.2.1.14
shrinking /nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14/lib/libtalloc-compat1-2.1.14.so
gzipping man pages under /nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14/share/man/
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14/lib
patching script interpreter paths in /nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
checking for references to /build in /nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14...
/nix/store/sfvr5ib3yqyxj2vzr832in3dsihqsriy-libxml2-2.9.8-bin
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
/nix/store/384xr2d5y2bshbyl28r7g20raj0r0j2l-unscd-0.53
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=c003f65a-8fec-404e-9930-2f95f06e73a1-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=8ac1981a-d329-4969-a321-ba21335bce50)

Attempted: libxml2, talloc

The following builds were skipped because they don't evaluate on x86_64-darwin: unscd

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14/lib/libpytalloc-util.2.1.14.dylib: fixing dylib
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14/lib/libtalloc-compat1-2.1.14.dylib: fixing dylib
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14/lib/libtalloc.2.1.14.dylib: fixing dylib
gzipping man pages under /nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14/share/man/
strip is /nix/store/g5r4apl0za012ffs6ladinwa5w0m1l3k-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14/lib
patching script interpreter paths in /nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
/nix/store/fy1lmal54cgkr714fv1iy3m5wprjks2h-libxml2-2.9.8-bin
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=9959f006-7373-400c-9a4a-9bc44d55d0e2-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=b2cb1203-f698-4683-98b0-c68b70f7291f)

Attempted: libxml2, qemu, talloc, vde2

The following builds were skipped because they don't evaluate on x86_64-darwin: unscd

<details><summary>Partial log (click to expand)</summary><p>

```
find: '/nix/store/k6fb10fq5k24zxrpyxd0dngxiam3vf7j-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/k6fb10fq5k24zxrpyxd0dngxiam3vf7j-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/k6fb10fq5k24zxrpyxd0dngxiam3vf7j-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/k6fb10fq5k24zxrpyxd0dngxiam3vf7j-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/k6fb10fq5k24zxrpyxd0dngxiam3vf7j-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/k6fb10fq5k24zxrpyxd0dngxiam3vf7j-qemu-3.0.0-ga': No such file or directory
/nix/store/fy1lmal54cgkr714fv1iy3m5wprjks2h-libxml2-2.9.8-bin
/nix/store/4fmxd0cc5ajjlxzna0yhixga3z93zi8l-qemu-3.0.0
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
/nix/store/kg78ci82c8z55qwgyazgvxwkcaf5v8y9-vde2-2.3.2
```
</p></details>

matthewbauer committed:
treewide: add emulator to platform

You can use stdenv.hostPlatform.emulator to get an executable that
runs cross-built binaries. This could be any emulator. For instance,
we use QEMU to emulate Linux targets and Wine to emulate Windows
targets. To work with qemu, we need to support custom targets.

I’ve reworked the cross tests in pkgs/test/cross to use this
functionality.

Also, I’ve used talloc to cross-execute with the emulator. There
appears to be a cross-execute for all waf builds. In the future, it
would be nice to set this for all waf builds.

Adds stdenv.hostPlatform.qemuArch attrbute to get the qemuArch for
each platform.

GrahamcOfBorg commented:
<!--REQUEST_ID=dbe22abd-3495-4c95-af2a-67eff844694e-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=e0fc47fe-e73c-4f50-9a66-f154f9379a71)

Attempted: libxml2, qemu, talloc, vde2

The following builds were skipped because they don't evaluate on x86_64-darwin: unscd

<details><summary>Partial log (click to expand)</summary><p>

```
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
/nix/store/fy1lmal54cgkr714fv1iy3m5wprjks2h-libxml2-2.9.8-bin
/nix/store/j7di4lkyb6hy7kk59nyhcfl7w0vvdzq1-qemu-3.0.0
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
/nix/store/kg78ci82c8z55qwgyazgvxwkcaf5v8y9-vde2-2.3.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=dbe22abd-3495-4c95-af2a-67eff844694e-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=aa983281-7fb1-45cb-96ae-d15e739496be)

Attempted: libxml2, qemu, talloc, unscd, vde2

<details><summary>Partial log (click to expand)</summary><p>

```
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
/nix/store/sfvr5ib3yqyxj2vzr832in3dsihqsriy-libxml2-2.9.8-bin
/nix/store/5bw1y54qnssfrjzj7sp8j48ygqiigyc5-qemu-3.0.0
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
/nix/store/384xr2d5y2bshbyl28r7g20raj0r0j2l-unscd-0.53
/nix/store/6s5w6k37xg6qkssn24mdwgb4mwxfsz1v-vde2-2.3.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=25acb1b5-3b31-45fa-8781-8442ff9fe4c6-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=9b01d4ac-cbf9-4889-8a22-ad7016f38850)

Attempted: libxml2, qemu, talloc, unscd, vde2

<details><summary>Partial log (click to expand)</summary><p>

```
these derivations will be built:
  /nix/store/4l7z42arkjbdhcvj1fxnms1gppyyaqgm-qemu-3.0.0.drv
waiting for locks or build slots...
/nix/store/sfvr5ib3yqyxj2vzr832in3dsihqsriy-libxml2-2.9.8-bin
/nix/store/5bw1y54qnssfrjzj7sp8j48ygqiigyc5-qemu-3.0.0
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
/nix/store/384xr2d5y2bshbyl28r7g20raj0r0j2l-unscd-0.53
/nix/store/6s5w6k37xg6qkssn24mdwgb4mwxfsz1v-vde2-2.3.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=dbe22abd-3495-4c95-af2a-67eff844694e-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=85304434-a7c7-4e21-97ee-b165bf6681bf)

Attempted: libxml2, qemu, talloc, unscd, vde2

<details><summary>Partial log (click to expand)</summary><p>

```
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
/nix/store/h73hfs1kkr9r8zr8rgjzivdc2ff0azdq-libxml2-2.9.8-bin
/nix/store/yqdyzx4l4jlwxawdcap4fqnfpmbsk6xh-qemu-3.0.0
/nix/store/h50z2qsy9mpl3swqp5pzwp0q6nblc83a-talloc-2.1.14
/nix/store/isbgw2kd3gqbrr05l24562zva9qssz57-unscd-0.53
/nix/store/rxpi1zsicn5dz1k0xvvj1gackbfydhgj-vde2-2.3.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=25acb1b5-3b31-45fa-8781-8442ff9fe4c6-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=91a9a20a-28e2-46d8-9f5f-6e63eebc5e32)

Attempted: libxml2, qemu, talloc, vde2

The following builds were skipped because they don't evaluate on x86_64-darwin: unscd

<details><summary>Partial log (click to expand)</summary><p>

```
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/bwlxphirlgkq6axjcbs08llzq5hygd31-qemu-3.0.0-ga': No such file or directory
/nix/store/fy1lmal54cgkr714fv1iy3m5wprjks2h-libxml2-2.9.8-bin
/nix/store/j7di4lkyb6hy7kk59nyhcfl7w0vvdzq1-qemu-3.0.0
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
/nix/store/kg78ci82c8z55qwgyazgvxwkcaf5v8y9-vde2-2.3.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=25acb1b5-3b31-45fa-8781-8442ff9fe4c6-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=8980369e-2893-47df-8464-c6d833e3594e)

Attempted: libxml2, qemu, talloc, unscd, vde2

<details><summary>Partial log (click to expand)</summary><p>

```
  CC      sparc-softmmu/accel/stubs/whpx-stub.o
  CC      sparc-softmmu/accel/stubs/kvm-stub.o
  CC      sparc-softmmu/accel/tcg/tcg-all.o
  CC      sparc-softmmu/accel/tcg/cputlb.o
  CC      sparc64-softmmu/tcg/tcg-op.o
  CC      sparc64-softmmu/tcg/tcg-op-vec.o
  CC      sparc-softmmu/accel/tcg/tcg-runtime.o
  CC      sparc-softmmu/accel/tcg/tcg-runtime-gvec.o
building of '/nix/store/rbsn6dx5dsar7zv3577n9lm2lv37mii1-qemu-3.0.0.drv' timed out after 1800 seconds
error: build of '/nix/store/rbsn6dx5dsar7zv3577n9lm2lv37mii1-qemu-3.0.0.drv' failed
```
</p></details>

matthewbauer commented:
@grahamcofborg build pkgsCross.aarch64-multiplatform.talloc pkgsCross.mingwW64.talloc pkgsCross.powernv.talloc tests.cross.aarch64-multiplatform tests.cross.mingwW64 tests.cross.musl64 tests.cross.fuloongminipc

GrahamcOfBorg commented:
<!--REQUEST_ID=1821d42d-71bd-4159-962b-20504cf46a43-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=d5d06f5d-1a2c-4f64-9921-2a254b87c11f)

Attempted: pkgsCross.aarch64-multiplatform.talloc, pkgsCross.powernv.talloc, tests.cross.aarch64-multiplatform, tests.cross.mingwW64, tests.cross.musl64, tests.cross.fuloongminipc

The following builds were skipped because they don't evaluate on x86_64-darwin: pkgsCross.mingwW64.talloc

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/jhy2fbyd1vb1c06g0ch116j0yzdjswqx-powerpc64le-unknown-linux-gnu-stage-final-gcc-debug-7.3.0.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/mjhswy85za756km5c986w8fwjqp3gbgk-powerpc64le-unknown-linux-gnu-stage-final-gcc-debug-wrapper-7.3.0.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/a9l4yhcd949kzpfcps7rzb49zpanhphj-stdenv-darwin.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/44fmr4idjlcyi3ckhx0wvxx11ahgfcr6-ncurses-6.1-powerpc64le-unknown-linux-gnu.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/fb4mq9cxf60sq7l2f004wrjfw2f9fny1-zlib-1.2.11-powerpc64le-unknown-linux-gnu.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/v902d570qvkcqs4z02wvi343ghqblix7-libxml2-2.9.8-powerpc64le-unknown-linux-gnu.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/bzrh7irsvf95w2py0riwp46a6gpjx37c-readline-6.3p08-powerpc64le-unknown-linux-gnu.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/kpff95hz3a548cz356vgyy616l2ih4xr-libxslt-1.1.32-powerpc64le-unknown-linux-gnu.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/iavzgcjrlfrn9pbr5l4ax2jypz51xraz-talloc-2.1.14-powerpc64le-unknown-linux-gnu.drv': 4 dependencies couldn't be built
error: build of '/nix/store/31xg5am39iwr8dr2anbjv6541nbbfxk9-test-file-5.34-x86_64-unknown-linux-musl.drv', '/nix/store/8zl1q7sdd7q4j6i1wr3sf1r14354bdqg-test-hello-2.10-mipsel-unknown-linux-gnu.drv', '/nix/store/905f9gigzz4awvjckbqa119b2cvcphpn-test-hello-2.10-x86_64-pc-mingw32.drv', '/nix/store/iavzgcjrlfrn9pbr5l4ax2jypz51xraz-talloc-2.1.14-powerpc64le-unknown-linux-gnu.drv', '/nix/store/jjp029zylglmkniai191y5nks22kjp4s-test-file-5.34-winlinks-x86_64-pc-mingw32.drv', '/nix/store/mlvwi3j68pi2lsqm9rgmwz0wg227y5x5-talloc-2.1.14-aarch64-unknown-linux-gnu.drv', '/nix/store/q6xqqy17864g80q0wm0cvkvwbcyb14ki-test-file-5.34-aarch64-unknown-linux-gnu.drv', '/nix/store/shpxls619jlqv1qb85fv8llpay6byf8n-test-file-5.34-mipsel-unknown-linux-gnu.drv', '/nix/store/y2mc8kx8vw4vyir1hd6id1gkjnhnpyjc-test-hello-2.10-aarch64-unknown-linux-gnu.drv', '/nix/store/yavw1gh54fl74qws5h0j5152lyvsgdrm-test-hello-2.10-x86_64-unknown-linux-musl.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=1821d42d-71bd-4159-962b-20504cf46a43-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=a91425b0-a823-4226-adba-d74e00249e84)

Attempted: pkgsCross.aarch64-multiplatform.talloc, pkgsCross.powernv.talloc, tests.cross.aarch64-multiplatform, tests.cross.mingwW64, tests.cross.musl64, tests.cross.fuloongminipc

The following builds were skipped because they don't evaluate on x86_64-linux: pkgsCross.mingwW64.talloc

<details><summary>Partial log (click to expand)</summary><p>

```
  ftmarshal.o git.o hglobalstream.o ifs.o itemmoniker.o marshal.o memlockbytes.o moniker.o ole2.o \
  ole2impl.o ole2stubs.o ole32_main.o oleobj.o oleproxy.o pointermoniker.o rpc.o stg_prop.o \
  stg_stream.o storage32.o stubmanager.o usrmarshal.o dlldata.o dcom_p.o irot_c.o ole32_objidl_p.o \
  ole32_oleidl_p.o ole32_unknwn_p.o ole32res.res dcom_r.res ole32_objidl_r.res ole32_oleidl_r.res \
  ole32_unknwn_r.res -loleaut32 ../../dlls/uuid/libuuid.a -ladvapi32 -luser32 -lgdi32 -lrpcrt4 \
  ../../libs/port/libwine_port.a -Wb,-doleaut32
building of '/nix/store/kagi726bcgrqc3c7sdj2bvxfi4187swc-wine64-3.0.3.drv' timed out after 1800 seconds
cannot build derivation '/nix/store/6n4yyjlf04c1brl1n9h38855sdkw9g64-test-file-5.34-winlinks-x86_64-pc-mingw32.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wg3w2abhqsbsnsny5103r4pw0yrdwqvj-test-hello-2.10-x86_64-pc-mingw32.drv': 1 dependencies couldn't be built
error: build of '/nix/store/36wzhc6zm9f53i0mw54wvygy4jns8d8m-talloc-2.1.14-powerpc64le-unknown-linux-gnu.drv', '/nix/store/4wd2vzydhfwhbzi1f5hp3rjmch8ybh6c-test-file-5.34-x86_64-unknown-linux-musl.drv', '/nix/store/6n4yyjlf04c1brl1n9h38855sdkw9g64-test-file-5.34-winlinks-x86_64-pc-mingw32.drv', '/nix/store/fkgnnri1ggwnihpci3axiw2ajv8w4zpc-test-hello-2.10-x86_64-unknown-linux-musl.drv', '/nix/store/wg3w2abhqsbsnsny5103r4pw0yrdwqvj-test-hello-2.10-x86_64-pc-mingw32.drv', '/nix/store/xaikgi8j965klw2h8b0iwi4z20c2wawn-test-hello-2.10-mipsel-unknown-linux-gnu.drv', '/nix/store/y4fzq7hz35gm4nqkzlzh88c0ghingdz9-test-file-5.34-mipsel-unknown-linux-gnu.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=384c0bef-8e4f-4c6f-9578-d51e4f2a5cab-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=e1bcf7ac-ac2e-4260-8ef4-45b7a6c4c7db)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /build in /nix/store/7iww4d7acalw05ikxbhprnyj15dlfsc9-qemu-3.0.0...
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
/nix/store/7iww4d7acalw05ikxbhprnyj15dlfsc9-qemu-3.0.0
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=384c0bef-8e4f-4c6f-9578-d51e4f2a5cab-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=a014c17b-c520-45ce-81b9-dd41559f6a81)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
find: '/nix/store/8ikj8qfg5dmmzja38pnag2ia4j7l33av-qemu-3.0.0-ga': No such file or directory
/nix/store/6wnyr7qpcihmi725548z7m2v6m68j51b-qemu-3.0.0
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=384c0bef-8e4f-4c6f-9578-d51e4f2a5cab-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=5c93b737-e723-43e1-9053-e8f324f5c252)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
  CC      mips64el-softmmu/accel/accel.o
  CC      mips64el-softmmu/accel/stubs/hax-stub.o
  CC      mips64el-softmmu/accel/stubs/hvf-stub.o
  CC      mips64el-softmmu/accel/stubs/whpx-stub.o
  CC      mips64el-softmmu/accel/stubs/kvm-stub.o
  CC      mips64el-softmmu/accel/tcg/tcg-all.o
  CC      mips64el-softmmu/accel/tcg/cputlb.o
  CC      mips64el-softmmu/accel/tcg/tcg-runtime.o
building of '/nix/store/kq26az0zzyl241d7nkz4870xsshr2722-qemu-3.0.0.drv' timed out after 3600 seconds
error: build of '/nix/store/kq26az0zzyl241d7nkz4870xsshr2722-qemu-3.0.0.drv' failed
```
</p></details>

matthewbauer committed:
systems/examples.nix: move riscv function to let binding

Makes it easier to use mapAttrs with lib.systems.examples. Now every
entry in it are legitimate systems.

GrahamcOfBorg commented:
<!--REQUEST_ID=12b1ac59-019f-4269-b63a-38cb4f06c598-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=9566da50-b53e-4471-9c63-e02308ae2f57)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/7iww4d7acalw05ikxbhprnyj15dlfsc9-qemu-3.0.0
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=12b1ac59-019f-4269-b63a-38cb4f06c598-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=803c9853-2d75-4dec-add1-62e24dfef0ee)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
  CC      s390x-softmmu/hw/s390x/tod-qemu.o
clang-5.0: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
  CC      s390x-softmmu/hw/s390x/s390-ccw.o
clang-5.0: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
  CC      s390x-softmmu/target/s390x/cpu.o
clang-5.0: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
  CC      s390x-softmmu/target/s390x/cpu_models.o
clang-5.0: warning: argument unused during compilation: '-fno-strict-overflow' [-Wunused-command-line-argument]
building of '/nix/store/prwbrdrj2j870lavhk8335bz0ijmxfcj-qemu-3.0.0.drv' timed out after 1800 seconds
error: build of '/nix/store/prwbrdrj2j870lavhk8335bz0ijmxfcj-qemu-3.0.0.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=12b1ac59-019f-4269-b63a-38cb4f06c598-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=8ee626d6-9e0e-48d2-af7a-3081add27dce)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
  CC      sparc64-softmmu/hw/scsi/vhost-user-scsi.o
  CC      tricore-softmmu/disas.o
  CC      sparc64-softmmu/hw/timer/mc146818rtc.o
  CC      tricore-softmmu/arch_init.o
  CC      sparc64-softmmu/hw/vfio/common.o
  CC      tricore-softmmu/cpus.o
  CC      sparc64-softmmu/hw/vfio/pci.o
  CC      tricore-softmmu/monitor.o
building of '/nix/store/kq26az0zzyl241d7nkz4870xsshr2722-qemu-3.0.0.drv' timed out after 1800 seconds
error: build of '/nix/store/kq26az0zzyl241d7nkz4870xsshr2722-qemu-3.0.0.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=7b9da6bf-8f26-47b9-90b6-7e8b99b5bce3-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=38c7047b-5c28-4a55-805a-3ea020af99d9)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/7iww4d7acalw05ikxbhprnyj15dlfsc9-qemu-3.0.0
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=7b9da6bf-8f26-47b9-90b6-7e8b99b5bce3-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=4a3e5940-b8d1-4af3-9de9-1139bd158947)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /build in /nix/store/vidhvq0b6gwa4nzyd622yfa4sb4nsqyb-qemu-3.0.0...
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
/nix/store/vidhvq0b6gwa4nzyd622yfa4sb4nsqyb-qemu-3.0.0
/nix/store/h50z2qsy9mpl3swqp5pzwp0q6nblc83a-talloc-2.1.14
```
</p></details>

matthewbauer commented:
@grahamcofborg build pkgs.tests.cross.hello

GrahamcOfBorg commented:
<!--REQUEST_ID=1b497f81-486c-4ba7-9a4f-c356fb503251-->
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=98b188f4-ac2a-4d2f-9c89-67cd195593cc)

The following builds were skipped because they don't evaluate on x86_64-darwin: pkgs.tests.cross.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgs.tests.cross.hello' because:
error: while evaluating the attribute 'buildCommand' of the derivation 'test-hello-2.10-aarch64-unknown-linux-android' at /private/var/lib/ofborg/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/lnl7-mac/pkgs/stdenv/generic/make-derivation.nix:180:11:
while evaluating anonymous function at /private/var/lib/ofborg/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/lnl7-mac/pkgs/build-support/exe-wrapper/default.nix:3:1, called from /private/var/lib/ofborg/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/lnl7-mac/pkgs/test/cross/default.nix:56:16:
Don't know how to run aarch64-unknown-linux-android executables.

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=1b497f81-486c-4ba7-9a4f-c356fb503251-->
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=c7bc7579-2ca0-4e6f-9572-711f56c75a9d)

The following builds were skipped because they don't evaluate on aarch64-linux: pkgs.tests.cross.hello

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

matthewbauer commented on lib/systems/default.nix:
```diff
@@ -66,6 +66,44 @@ rec {
+          "powerpc64le" = "ppc64";
+        }.${final.parsed.cpu.name} or final.parsed.cpu.name;
+
+      emulator = pkgs: let
```
@ericson2314 How do you feel about putting this in lib/systems/default.nix? I want to avoid sticking it in "pkgs" as that is only really supposed to contain callPackage usages.

Mic92 replied:
Can you move the emulator stuff to its own file?

GrahamcOfBorg commented:
<!--REQUEST_ID=2341d904-fd76-4eb2-a057-9585e64ba18e-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=e4af9657-1e1e-4b2e-bab1-ddca04b5ab28)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /build in /nix/store/7iww4d7acalw05ikxbhprnyj15dlfsc9-qemu-3.0.0...
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
/nix/store/7iww4d7acalw05ikxbhprnyj15dlfsc9-qemu-3.0.0
/nix/store/0hnlzq0rhnbiz6j2f1yjx3hhv91hx3x0-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=2341d904-fd76-4eb2-a057-9585e64ba18e-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=b524aacd-98b5-4f75-8cd1-7270e8a04978)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/6wnyr7qpcihmi725548z7m2v6m68j51b-qemu-3.0.0
/nix/store/aaq57h07ph8fxj1aqlwhzpwhszwqjmzr-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=2341d904-fd76-4eb2-a057-9585e64ba18e-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=53235515-59db-4188-a5de-f25b5e507b9f)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
  CC      mips64el-softmmu/accel/stubs/hax-stub.o
  CC      mips64el-softmmu/accel/stubs/hvf-stub.o
  CC      mips64el-softmmu/accel/stubs/whpx-stub.o
  CC      mips64el-softmmu/accel/stubs/kvm-stub.o
  CC      mips64el-softmmu/accel/tcg/tcg-all.o
  CC      mips64el-softmmu/accel/tcg/cputlb.o
  CC      mips64el-softmmu/accel/tcg/tcg-runtime.o
  CC      mips64el-softmmu/accel/tcg/tcg-runtime-gvec.o
building of '/nix/store/7np6im1syiv7b10rjfih0q5fzqpirlnh-qemu-3.0.0.drv' timed out after 3600 seconds
error: build of '/nix/store/7np6im1syiv7b10rjfih0q5fzqpirlnh-qemu-3.0.0.drv' failed
```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build pkgs.tests.cross.hello.aarch64-multiplatform

GrahamcOfBorg commented:
<!--REQUEST_ID=bcd0db32-4d69-4d2a-81f7-c1f7abf1129e-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=ea8d1f51-9749-45e6-b003-697ad65e4a64)

Attempted: pkgs.tests.cross.hello.aarch64-multiplatform

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/pxsv55dmnybiw3vlj09xm9wqp1sjvzaa-aarch64-unknown-linux-gnu-stage-static-gcc-debug-wrapper-7.3.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/2xxlwhl4xbrhap9vcbznqb0dc499amgd-stdenv-linux.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ivsl8nn7wchlk18mdlvpvi4x4h4l9z2z-glibc-2.27-aarch64-unknown-linux-gnu.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/27q3wv9ickn1jzwl56hhs5qzymc0bimh-aarch64-unknown-linux-gnu-binutils-wrapper-2.30.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/mfii6k3b9gkzx3di873igc4zv25lm486-aarch64-unknown-linux-gnu-stage-final-gcc-debug-7.3.0.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/y1pmi8r81h54c1ywwwzvynrg4i29lw94-aarch64-unknown-linux-gnu-stage-final-gcc-debug-wrapper-7.3.0.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/slr4vxyf1lx8gvi4gykl9n6lwpzqlg5v-stdenv-linux.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/idwl0np1x4bqy78hf0rygfx258dx7wik-hello-2.10-aarch64-unknown-linux-gnu.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/16qfbkdq7dnlfk2713nscwnckkp07p6i-test-hello-2.10-aarch64-unknown-linux-gnu.drv': 1 dependencies couldn't be built
error: build of '/nix/store/16qfbkdq7dnlfk2713nscwnckkp07p6i-test-hello-2.10-aarch64-unknown-linux-gnu.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=bcd0db32-4d69-4d2a-81f7-c1f7abf1129e-->
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=159d99f1-35ea-43b0-9c94-730d87363474)

The following builds were skipped because they don't evaluate on x86_64-linux: pkgs.tests.cross.hello.aarch64-multiplatform

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgs.tests.cross.hello.aarch64-multiplatform' because:
error: while evaluating the attribute 'buildCommand' of the derivation 'test-hello-2.10-aarch64-unknown-linux-gnu' at /var/lib/gc-of-borg/.nix-test-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/rbvermaa-spot/pkgs/stdenv/generic/make-derivation.nix:180:11:
while evaluating 'emulator' at /var/lib/gc-of-borg/.nix-test-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/rbvermaa-spot/lib/systems/default.nix:80:18, called from /var/lib/gc-of-borg/.nix-test-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/rbvermaa-spot/pkgs/test/cross/default.nix:56:16:
while evaluating the attribute 'platform.isi686' at /var/lib/gc-of-borg/.nix-test-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/rbvermaa-spot/lib/systems/examples.nix:44:5:
attribute 'isi686' missing, at /var/lib/gc-of-borg/.nix-test-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/rbvermaa-spot/lib/systems/default.nix:99:14

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=9820dd79-7111-42a2-8641-9f170454f97e-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=afc1c03b-7c86-4cb0-a848-867eb06c40d6)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/n4hb93w6j076xcjw5pm09rdmc09s075b-binutils-2.30/bin/strip
cannot link '/nix/store/.links/1l5qq55ff1z51vg9xhfwn6z0cjz5vdvg40fz1675js55impf4krq' to '/nix/store/90rf7n73slqy9db4ly797qh5psfzwgl0-qemu-3.0.0/bin/qemu-ppc64': No space left on device
/nix/store/90rf7n73slqy9db4ly797qh5psfzwgl0-qemu-3.0.0
/nix/store/n56ax0ynjv31k0bxv46las6dmvgfg9ls-talloc-2.1.14
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=9820dd79-7111-42a2-8641-9f170454f97e-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.50212&attempt_id=1ce9a66d-dcfa-4ada-bc4c-8f5b4dbcf0ad)

Attempted: qemu, talloc

<details><summary>Partial log (click to expand)</summary><p>

```
checking for references to /build in /nix/store/5x6l8rqnyiing1xw9jli9i1xlp37fmm6-qemu-3.0.0...
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/qjrnv0qw44bw1hc23zhfh26xd1c25dfs-binutils-2.30/bin/strip
/nix/store/5x6l8rqnyiing1xw9jli9i1xlp37fmm6-qemu-3.0.0
/nix/store/frpy0161swf0aksaz1nsry4v10clxfks-talloc-2.1.14
```
</p></details>

lopsided98 commented:
I wonder if this can be used to help cross compile packages that use gobject-introspection.
